engine/api/v2_workflow_run_engine.go

```diff
@@ -0,0 +1,384 @@
+			return err
+		}
+	}
+
+	// Save Run message
+	for i := range runMsgs {
+		if err := workflow_v2.InsertRunInfo(ctx, tx, &runMsgs[i]); err != nil {
+			return err
+		}
+	}
+
+	// End workflow if there is no job to queue,  no running jobs and current status is not terminated
+	if len(jobsToQueue) == 0 && sdk.StatusIsTerminated(currentJobRunStatus) && !sdk.StatusIsTerminated(run.Status) {
+		run.Status = currentJobRunStatus
+
+		if err := workflow_v2.UpdateRun(ctx, tx, run); err != nil {
+			return err
+		}
+	}
+
+	return sdk.WithStack(tx.Commit())
+}
+
+// TODO manage re run
+func retrieveJobToQueue(ctx context.Context, db *gorp.DbMap, run *sdk.V2WorkflowRun, wrEnqueue sdk.V2WorkflowRunEnqueue, u *sdk.AuthentifiedUser, defaultRegion string) (map[string]sdk.V2Job, []sdk.V2WorkflowRunInfo, string, error) {
+	runInfos := make([]sdk.V2WorkflowRunInfo, 0)
+	jobToQueue := make(map[string]sdk.V2Job)
+
+	// Load run_jobs
+	_, next := telemetry.Span(ctx, "retrieveJobToQueue.LoadRunJobsByRunID")
+	runJobs, err := workflow_v2.LoadRunJobsByRunID(ctx, db, run.ID)
+	if err != nil {
+		next()
+		return nil, nil, "", sdk.WrapError(err, "unable to load workflow run jobs for run %s", wrEnqueue.RunID)
+	}
+	next()
+
+	// If jobs has already been queue and wr not terminated and user want to trigger some job => error
+	if !sdk.StatusIsTerminated(run.Status) && len(wrEnqueue.Jobs) > 0 && len(runJobs) > 0 {
+		info := sdk.V2WorkflowRunInfo{
+			WorkflowRunID: run.ID,
+			Level:         sdk.WorkflowRunInfoLevelWarning,
+			Message:       "unable to start a job on a running workflow",
+		}
+		runInfos = append(runInfos, info)
+		return nil, runInfos, "", nil
+	}
+
+	// Compute run context
+	jobsContext := buildJobsContext(runJobs)
+
+	// Select jobs to check ( all workflow or list of jobs from enqueue request )
+	jobsToCheck := make(map[string]sdk.V2Job)
+	if len(wrEnqueue.Jobs) == 0 {
+		for jobID, jobDef := range run.WorkflowData.Workflow.Jobs {
+			jobsToCheck[jobID] = jobDef
+		}
+	} else {
+		for _, jobID := range wrEnqueue.Jobs {
+			if jobDef, has := run.WorkflowData.Workflow.Jobs[jobID]; has {
+				jobsToCheck[jobID] = jobDef
+			}
+		}
+	}
+
+	// Check jobs : Needs / Condition / User Right
+	for jobID, jobDef := range jobsToCheck {
+
+		// TODO manage re run
+		if _, has := jobsContext[jobID]; has {
+			log.Debug(ctx, "job %s: already executed, skip it", jobID)
+			continue
+		}
+
+		// Check jobs needs
+		requiredJob, ok := checkJobNeeds(jobsContext, jobDef)
+		if !ok {
+			// If not ok , and ask to run it => send message
+			if len(wrEnqueue.Jobs) > 0 {
+				runInfos = append(runInfos, sdk.V2WorkflowRunInfo{
+					WorkflowRunID: run.ID,
+					Level:         sdk.WorkflowRunInfoLevelWarning,
+					Message:       fmt.Sprintf("job %s: missing some required job: %s", jobID, requiredJob),
+				})
+			}
+			continue
+		}
+
+		_, next = telemetry.Span(ctx, "retrieveJobToQueue.checkUserRight")
+		hasRight, err := checkUserRight(ctx, db, jobDef, *u, defaultRegion)
```

Comment:
it's probably not the right user we are checking here

Reply:
We are checking the right user, the one that enqueue the event to run the workflow.